INSTALL

```diff
@@ -463,6 +463,14 @@ MFEM_USE_HIOP = YES/NO
    Enable the usage of HiOp (https://github.com/LLNL/hiop) in MFEM. HiOp is an
    HPC solver for nonlinear optimization problems.
 
+MFEM_USE_CODIPACK = YES/NO
+  Enable automatic differentiation using the CoDiPack library.
+  www.scicomp.uni-kl.de/codi/
+
+MFEM_USE_ADFORWARD = YES/NO
+  Enable forward mode for AD packages. This option is valid
+  only if the AD package supports two modes (backward/forward).
```

Comment:
The documentation should make it clear that choice one is between CODIPACK and an internal AD implementation. The second choice is forward mode versus reverse mode AD. It would be useful to explain when it makes sense to make one choice versus the other. Presumably, there are pluses and minuses to each choice. It would be nice to have those pluses and minuses summarized. If there is a safe default, that's also worth mentioning.

Reply:
MFEM_USE_ADFORWARD can be YES if MFEM_USE_CODIPACK is NO
For native implementation it doesn't have any effect, however, I am planning to add two additional AD libraries and MFEM_USE_ADFORWARD will affect them. I am writing a paper on what, when and how should be used - once it is ready I will add it to the documentation. 

If CODIPACK is not ON - the native implementation will be automatically enabled. So far the numerical experiments with the examples did not show any advantages of CODIPACK vs the native implementation. Clearly, if one needs reverse mode (scalar function with large number of arguments) it will be preferable to use CODIPACK. However, presenting the full AD theory in the code is a bit too much. The users will be referred to the coming paper.